File: CodeGenDotNet 159059-1372 (a)/VBCode/Chapter 1/HelloWorldAndSimpleClass/ClassViaCodeDOM.vb
Option Strict On
Option Explicit On

Imports System

'! Class Summary: Create Hello World program using the CodeDOM

Public Class ClassViaCodeDOM

#Region "Public Methods and Properties"
   Public Shared Function BuildGraph( _
               ByVal xmlMetaData As Xml.XmlDocument, _
               ByVal tableName As String) _
               As CodeDom.CodeCompileUnit
      Dim node As Xml.XmlNode
      Dim nodeList As Xml.XmlNodeList
      Dim compileUnit As New CodeDom.CodeCompileUnit
      Dim nSpace As CodeDom.CodeNamespace
      Dim clsTable As CodeDom.CodeTypeDeclaration

      nodeList = xmlMetaData.SelectNodes("/DataSet/Table[@Name='" & tableName & "']/Column")
      compileUnit.UserData.Add("AllowLateBound", False)
      compileUnit.UserData.Add("RequireVariableDeclaration", True)

      nSpace = New CodeDom.CodeNamespace("ClassViaCodeDOM")
      compileUnit.Namespaces.Add(nSpace)

      nSpace.Imports.Add(New CodeDom.CodeNamespaceImport("System"))

      clsTable = New CodeDom.CodeTypeDeclaration(tableName)
      nSpace.Types.Add(clsTable)

      Dim field As CodeDom.CodeMemberField
      For Each node In nodeList
         field = New CodeDom.CodeMemberField
         field.Name = "m_" & node.Attributes("Name").Value
         field.Attributes = CodeDom.MemberAttributes.Private
         field.Type = New CodeDom.CodeTypeReference(node.Attributes("Type").Value)
         clsTable.Members.Add(field)
      Next

      Dim prop As CodeDom.CodeMemberProperty
      Dim statement As CodeDom.CodePropertySetValueReferenceExpression
      Dim Name As String
      For Each node In nodeList
         prop = New CodeDom.CodeMemberProperty
         Name = node.Attributes("Name").Value
         prop.Name = Name
         prop.Attributes = CodeDom.MemberAttributes.Public
         prop.Type = New CodeDom.CodeTypeReference(node.Attributes("Type").Value)
         prop.GetStatements.Add(New CodeDom.CodeMethodReturnStatement(New CodeDom.CodeFieldReferenceExpression(New CodeDom.CodeThisReferenceExpression, "m_" & Name)))
         prop.SetStatements.Add(New CodeDom.CodeAssignStatement(New CodeDom.CodeFieldReferenceExpression(New CodeDom.CodeThisReferenceExpression, "m_" & Name), New CodeDom.CodePropertySetValueReferenceExpression))
         clsTable.Members.Add(prop)
      Next

      Return compileUnit
   End Function

   Public Shared Sub GenerateCode( _
               ByVal fileName As String, _
               ByVal provider As CodeDom.Compiler.CodeDomProvider, _
               ByVal compileunit As CodeDom.CodeCompileUnit)
      Dim gen As CodeDom.Compiler.ICodeGenerator = provider.CreateGenerator()
      Dim tw As CodeDom.Compiler.IndentedTextWriter
      Try
         tw = New CodeDom.Compiler.IndentedTextWriter(New IO.StreamWriter(fileName, False), "    ")
         gen.GenerateCodeFromCompileUnit(compileunit, tw, New CodeDom.Compiler.CodeGeneratorOptions)
      Finally
         If Not tw Is Nothing Then
            tw.Flush()
            tw.Close()
         End If
      End Try
   End Sub
#End Region

#Region "Protected and Friend Methods and Properties - empty"
#End Region

#Region "Private Methods and Properties - empty"
#End Region

End Class

File: CodeGenDotNet 159059-1372 (a)/CSharp/Chapter 1/HelloWorldAndSimpleClass/ClassViaCodeDOM.cs
using System;

/// <summary>
/// Create Hello World program using the CodeDOM
/// </summary>
public class ClassViaCodeDOM
{
	#region Public Methods and Properties
	public static System.CodeDom.CodeCompileUnit BuildGraph(	System.Xml.XmlDocument xmlMetaData,
																string tableName )
	{
		System.Xml.XmlNodeList nodeList;
		System.CodeDom.CodeCompileUnit compileUnit = new System.CodeDom.CodeCompileUnit();
		System.CodeDom.CodeNamespace nSpace;
		System.CodeDom.CodeTypeDeclaration clsTable;
		
		nodeList = xmlMetaData.SelectNodes( "/DataSet/Table[@Name='" + tableName + "']/Column" );

		nSpace = new System.CodeDom.CodeNamespace( "ClassViaCodeDOM" );
		compileUnit.Namespaces.Add( nSpace );
		
		nSpace.Imports.Add( new System.CodeDom.CodeNamespaceImport( "System" ) );
		
		clsTable = new System.CodeDom.CodeTypeDeclaration( tableName );
		nSpace.Types.Add( clsTable );

		System.CodeDom.CodeMemberField field;
		foreach( System.Xml.XmlNode node in nodeList )
		{
			field = new System.CodeDom.CodeMemberField();
			field.Name = "m_" + node.Attributes["Name"].Value;
			field.Attributes = System.CodeDom.MemberAttributes.Private;
			field.Type = new System.CodeDom.CodeTypeReference( node.Attributes["Type"].Value );
			clsTable.Members.Add( field );
		}
		
		System.CodeDom.CodeMemberProperty prop;
		string name;
		foreach( System.Xml.XmlNode node in nodeList )
		{
			prop = new System.CodeDom.CodeMemberProperty();
			name = node.Attributes["Name"].Value;
			prop.Name = name;
			prop.Attributes = System.CodeDom.MemberAttributes.Public;
			prop.Type = new System.CodeDom.CodeTypeReference( node.Attributes["Type"].Value );
			prop.GetStatements.Add( new System.CodeDom.CodeMethodReturnStatement( new System.CodeDom.CodeFieldReferenceExpression( new System.CodeDom.CodeThisReferenceExpression(), "m_" + name ) ) );
			prop.SetStatements.Add( new System.CodeDom.CodeAssignStatement( new System.CodeDom.CodeFieldReferenceExpression( new System.CodeDom.CodeThisReferenceExpression(), "m_" + name ), new System.CodeDom.CodePropertySetValueReferenceExpression() ) );
			clsTable.Members.Add( prop );
		}
		return compileUnit;
	}

	public static void GenerateCode(	string fileName, 
										System.CodeDom.Compiler.CodeDomProvider provider,
										System.CodeDom.CodeCompileUnit compileUnit )
	{
		System.CodeDom.Compiler.ICodeGenerator gen = provider.CreateGenerator();
		System.CodeDom.Compiler.IndentedTextWriter tw = null;
		try
		{
			tw = new System.CodeDom.Compiler.IndentedTextWriter( new System.IO.StreamWriter( fileName, false ), "	" );
			gen.GenerateCodeFromCompileUnit( compileUnit, tw, new System.CodeDom.Compiler.CodeGeneratorOptions() );
		}
		finally
		{
			if( tw != null )
			{
				tw.Flush();
				tw.Close();
			}
		}
	}
	#endregion

	#region Protected and Friend Methods and Properties - empty
	#endregion

	#region Private Methods and Properties - empty
	#endregion
}
